From 08b96217d3dbd31bf062129a8bedf15c5357f3e1 Mon Sep 17 00:00:00 2001
Message-ID: <08b96217d3dbd31bf062129a8bedf15c5357f3e1.1783491438.git.robux4@ycbcr.xyz>
In-Reply-To: <1e4ed1d4abbcfbf02ef80e000ce8ebd841dcf667.1783491438.git.robux4@ycbcr.xyz>
References: <1e4ed1d4abbcfbf02ef80e000ce8ebd841dcf667.1783491438.git.robux4@ycbcr.xyz>
From: Thomas Guillem <thomas@gllm.fr>
Date: Thu, 16 Jun 2022 14:21:36 +0200
Subject: [PATCH 15/20] modules: access: handle vlc_credential_get interruption

Fix a possible interrupt miss if vlc_credential_get() receive an
interruption and if access don't check vlc_killed() before doing I/O.

Refs #27027

(cherry picked from commit c55033cf75e9145bb58b81a2c3ef6bd55aa98c1b)
Signed-off-by: Thomas Guillem <thomas@gllm.fr>
Message-Id: <6176154d0c9208796d2553d9c37ae9f1c06d132a.1657539375.git.thomas@gllm.fr>
Message-Id: <4a03fdfa07263746eac019279ddcd2636ac2cc85.1662737834.git.dev.asenat@posteo.net>
Message-Id: <293b92570d648d85bad155b8a0fbbf5183bf23bb.1664884159.git.dev.asenat@posteo.net>
Message-Id: <f2f0f705190697375d9d62c92a472d0d4bab7ac8.1664886558.git.dev.asenat@posteo.net>
Message-Id: <50ba0aa698fd6235754bcba51cd601a308295f7e.1738244316.git.dcn.mcnamara@gmail.com>
---
 modules/access/dsm/access.c                    |  2 ++
 modules/access/ftp.c                           |  3 +++
 modules/access/http.c                          |  2 ++
 modules/access/live555.cpp                     |  5 +++++
 modules/access/sftp.c                          |  4 ++++
 modules/access/smb.c                           | 10 ++++++++++
 modules/access/smb2.c                          |  8 ++++++++
 modules/demux/adaptive/http/HTTPConnection.cpp |  7 +++++++
 8 files changed, 41 insertions(+)

diff --git a/modules/access/dsm/access.c b/modules/access/dsm/access.c
index fcf5625ca7..ebacf1c187 100644
--- a/modules/access/dsm/access.c
+++ b/modules/access/dsm/access.c
@@ -453,6 +453,8 @@ static int login( stream_t *p_access )
 
     vlc_credential_get( &credential, p_access, "smb-user", "smb-pwd",
                         NULL, NULL );
+    if (vlc_killed())
+        goto error;
 
     if( !credential.psz_username )
     {
diff --git a/modules/access/ftp.c b/modules/access/ftp.c
index 66170e613e..6399f43a03 100644
--- a/modules/access/ftp.c
+++ b/modules/access/ftp.c
@@ -455,6 +455,9 @@ static int Login( vlc_object_t *p_access, access_sys_t *p_sys, const char *path
     /* First: try credentials from url / option */
     vlc_credential_get( &credential, p_access, "ftp-user", "ftp-pwd",
                         NULL, NULL );
+    if (vlc_killed())
+        goto error;
+
     do
     {
         const char *psz_username = credential.psz_username;
diff --git a/modules/access/http.c b/modules/access/http.c
index 99d386c98e..59b8487515 100644
--- a/modules/access/http.c
+++ b/modules/access/http.c
@@ -269,6 +269,8 @@ static int Open( vlc_object_t *p_this )
         p_sys->url.psz_username = (char *) credential.psz_username;
         p_sys->url.psz_password = (char *) credential.psz_password;
     }
+    else if (vlc_killed())
+        goto error;
 
 connect:
     /* Connect */
diff --git a/modules/access/live555.cpp b/modules/access/live555.cpp
index a495187763..8d34c106ad 100644
--- a/modules/access/live555.cpp
+++ b/modules/access/live555.cpp
@@ -625,6 +625,11 @@ static int Connect( demux_t *p_demux )
         psz_user = credential.psz_username;
         psz_pwd = credential.psz_password;
     }
+    else if (vlc_killed())
+    {
+        i_ret = VLC_EGENERIC;
+        goto bailout;
+    }
 
 createnew:
     /* FIXME: This is naive and incorrect; it does not prevent the thread
diff --git a/modules/access/sftp.c b/modules/access/sftp.c
index 5af454239a..4cc3737fce 100644
--- a/modules/access/sftp.c
+++ b/modules/access/sftp.c
@@ -39,6 +39,7 @@
 #include <vlc_network.h>
 #include <vlc_url.h>
 #include <vlc_keystore.h>
+#include <vlc_interrupt.h>
 
 #include <libssh2.h>
 #include <libssh2_sftp.h>
@@ -400,6 +401,9 @@ static int Open( vlc_object_t* p_this )
 
     vlc_credential_get( &credential, p_access, "sftp-user", "sftp-pwd",
                         NULL, NULL );
+    if (vlc_killed())
+        goto error;
+
     char* psz_userauthlist = NULL;
     bool b_publickey_tried = false;
     do
diff --git a/modules/access/smb.c b/modules/access/smb.c
index 6add2a3e62..5086d0f733 100644
--- a/modules/access/smb.c
+++ b/modules/access/smb.c
@@ -52,6 +52,7 @@
 #include <vlc_input_item.h>
 #include <vlc_url.h>
 #include <vlc_keystore.h>
+#include <vlc_interrupt.h>
 #include <vlc_charset.h>
 
 #include "smb_common.h"
@@ -195,6 +196,15 @@ static int Open( vlc_object_t *p_this )
     credential.psz_realm = psz_var_domain;
     vlc_credential_get( &credential, p_access, "smb-user", "smb-pwd",
                         NULL, NULL );
+    if (vlc_killed())
+    {
+        vlc_credential_clean( &credential );
+        free(psz_var_domain);
+        free( psz_decoded_path );
+        vlc_UrlClean( &url );
+        return VLC_ENOMEM;
+    }
+
     for (;;)
     {
         if( smb_get_uri( p_access, &psz_uri, credential.psz_realm,
diff --git a/modules/access/smb2.c b/modules/access/smb2.c
index 6254fd84e2..23dbc6bcc6 100644
--- a/modules/access/smb2.c
+++ b/modules/access/smb2.c
@@ -884,6 +884,14 @@ Open(vlc_object_t *p_obj)
      * keystore/user interaction) */
     vlc_credential_get(&credential, access, "smb-user", "smb-pwd", NULL,
                        NULL);
+    if (vlc_killed())
+    {
+        vlc_credential_clean(&credential);
+        free(resolved_host);
+        ret = -EINTR;
+        goto error;
+
+    }
     ret = vlc_smb2_connect_open_share(access, url, &credential, false);
     if (ret == -EINVAL && credential.psz_username == NULL)
     {
diff --git a/modules/demux/adaptive/http/HTTPConnection.cpp b/modules/demux/adaptive/http/HTTPConnection.cpp
index 9e4c94a315..67cc54c31a 100644
--- a/modules/demux/adaptive/http/HTTPConnection.cpp
+++ b/modules/demux/adaptive/http/HTTPConnection.cpp
@@ -29,6 +29,7 @@
 
 #include <vlc_stream.h>
 #include <vlc_keystore.h>
+#include <vlc_interrupt.h>
 
 #include <new>
 
@@ -318,6 +319,12 @@ RequestStatus LibVLCHTTPConnection::request(const std::string &path,
         vlc_http_res_set_login(source->http_res,
                                crd.psz_username, crd.psz_password);
     }
+    else if (vlc_killed())
+    {
+        vlc_credential_clean(&crd);
+        vlc_UrlClean(&crd_url);
+        return RequestStatus::GenericError;
+    }
 
     int status = vlc_http_res_get_status(source->http_res);
     if (status < 0)
-- 
2.51.0

